### ISSUE TITLE
Liquibase is very slow on startup with spring boot

### ISSUE THREAD

**Issue by ggdn** [liquibase/liquibase#1312]
## Environment
**Liquibase Version**: 3.8.5
**Liquibase Integration & Version**: spring boot
**Liquibase Extension(s) & Version**: none
**Database Vendor & Version**: oracle v12
**Operating System Type & Version**: unix rhel v7

## Description
Hello 
Sometime Liquibase can take a lot of time to startup ~3 minutes without any changelog to apply. I don't know why, and the problem didn't occurred with the last version i used 3.6.3. I run Liquibase in spring boot applications.

## Steps To Reproduce
I don't know how to reproduce it. It appends randomly 

## Screenshots (if appropriate)
```
2020-08-09 06:40:16.956 [main] INFO  o.h.annotations.common.Version.<clinit> - HCANN000001: Hibernate Commons Annotations {5.1.0.Final}
2020-08-09 06:40:16.763 [main] INFO  org.hibernate.Version.logVersion - HHH000412: Hibernate Core {5.4.10.Final}
2020-08-09 06:40:16.677 [main] INFO  o.h.jpa.internal.util.LogHelper.logPersistenceUnitInformation - HHH000204: Processing PersistenceUnitInfo [name: default]
2020-08-09 06:40:16.558 [main] INFO  l.lockservice.StandardLockService.info - Successfully released change log lock
2020-08-09 06:40:16.493 [main] INFO  liquibase.executor.jvm.JdbcExecutor.info - SELECT * FROM DATABASECHANGELOG ORDER BY DATEEXECUTED ASC, ORDEREXECUTED ASC
2020-08-09 06:40:16.492 [main] INFO  l.c.StandardChangeLogHistoryService.info - Reading from DATABASECHANGELOG
2020-08-09 06:40:16.491 [main] INFO  liquibase.executor.jvm.JdbcExecutor.info - SELECT COUNT(*) FROM DATABASECHANGELOG
2020-08-09 06:40:16.483 [main] INFO  liquibase.executor.jvm.JdbcExecutor.info - SELECT MD5SUM FROM DATABASECHANGELOG WHERE MD5SUM IS NOT NULL AND ROWNUM=1
2020-08-09 06:35:51.224 [main] INFO  l.lockservice.StandardLockService.info - Successfully acquired change log lock
2020-08-09 06:35:51.207 [main] INFO  liquibase.executor.jvm.JdbcExecutor.info - SELECT LOCKED FROM DATABASECHANGELOGLOCK WHERE ID=1 FOR UPDATE
2020-08-09 06:35:51.205 [main] INFO  liquibase.executor.jvm.JdbcExecutor.info - SELECT COUNT(*) FROM DATABASECHANGELOGLOCK
2020-08-09 06:35:51.198 [main] INFO  liquibase.executor.jvm.JdbcExecutor.info - SELECT COUNT(*) FROM DATABASECHANGELOGLOCK
2020-08-09 06:35:50.684 [main] INFO  l.database.core.OracleDatabase.info - Could not set remarks reporting on OracleDatabase: com.zaxxer.hikari.pool.HikariProxyConnection.setRemarksReporting(boolean)
```

**Comment by molivasdat** [comment]
Hi @ggdn Thanks for writing up this issue.  From the log you shared, are there multiple springboot applications connecting to the same database?  Are there multiple changelogs trying to update the database at the same time?
I'm trying to figure out how to reproduce the error.  Random issues are difficult to debug and fix as you know.

**Comment by ggdn** [comment]
Hi thanks for your support. 
Yes, multiples applications are connecting to the same database, but they have there own dedicate schema 
Each schema/application has it own DATABASECHANGELOGLOCK/DATABASECHANGELOG.

All changelogs are already done on a previous startup

**Comment by ggdn** [comment]
I ran another test with debug log level. 1m30s between two lines :

```
2020-08-11 16:45:57.469 [main] INFO  liquibase.executor.jvm.JdbcExecutor.info - SELECT MD5SUM FROM REPORT_DEV.DATABASECHANGELOG WHERE MD5SUM IS NOT NULL AND ROWNUM=1
2020-08-11 16:44:37.208 [main] DEBUG liquibase.util.MD5Util.debug - Computed checksum for 1597157077208 as 36c3cc38b0b2d01ba88cb6caba17d9ca
```

all changelogs were already executed in another deployment
```
select DATEEXECUTED, EXECTYPE, MD5SUM, DESCRIPTION, LIQUIBASE from DATABASECHANGELOG ORDER BY ORDEREXECUTED;
2019-11-27 15:09:47.827433	EXECUTED	8:92a874f250a43dc3c3b0e425b7b1d0a5	tagDatabase	3.6.3
2019-11-27 15:09:48.182279	EXECUTED	8:ee5314380c0e1a9d9887c9a50e2ff1d8	sqlFile	3.6.3
2019-11-27 15:09:48.190352	EXECUTED	8:d69a04a7a9a06d5434f365177cee26c0	tagDatabase	3.6.3
2020-02-06 15:19:35.321187	EXECUTED	8:f4b6237bc82da01064875ccc8b2945cc	tagDatabase	3.8.2
2020-02-06 15:19:35.374909	EXECUTED	8:42a77bb4a9a345f056bdd4df6a394962	sqlFile	3.8.2
2020-02-06 15:19:35.386448	EXECUTED	8:9989df30ce77881207206dd97ca66e8b	tagDatabase	3.8.2
2020-08-07 16:04:01.413517	EXECUTED	8:af38df33b3d898fafed58f235af47c67	tagDatabase	3.8.5
2020-08-07 16:11:08.873591	EXECUTED	8:44e072bde80df172c2d0a8c976a5bd19	sqlFile	3.8.5
2020-08-07 16:11:08.882042	EXECUTED	8:ba38a141866d04978eb3fedcc757080b	tagDatabase	3.8.5
```

**Comment by molivasdat** [comment]
@ggdn Thanks for the additional info. Still puzzling on why it's getting stuck there.  Will need to look into the code between those two code sections.

**Comment by ggdn** [comment]
After investigation with my DBAs, we found that after 3.6.3, liquibase executes this query 9 times:
```
SELECT  NULL AS TABLE_CAT, atc.OWNER AS TABLE_SCHEMA, atc.OWNER, 
  atc.TABLE_NAME, atc.COLUMN_NAME, NULLABLE, ac.VALIDATED as VALIDATED, 
  ac.SEARCH_CONDITION, ac.CONSTRAINT_NAME 
FROM
 ALL_TAB_COLS atc JOIN all_cons_columns acc ON atc.OWNER = acc.OWNER AND 
  atc.TABLE_NAME = acc.TABLE_NAME AND atc.COLUMN_NAME = acc.COLUMN_NAME JOIN 
  all_constraints ac ON atc.OWNER = ac.OWNER AND atc.TABLE_NAME = 
  ac.TABLE_NAME AND acc.CONSTRAINT_NAME = ac.CONSTRAINT_NAME  WHERE atc.OWNER=
  'MY_SCHEMA' AND atc.hidden_column='NO' AND ac.CONSTRAINT_TYPE='C'  and 
  ac.search_condition is not null  AND atc.TABLE_NAME='DATABASECHANGELOG'
```

When the parsing of this query is not in cache, Oracle will find 5 executions plans (18,5 seconds each plan = 1minutes 30 secondes)
When the parsing is in cache, Oracle will use only one and unique execution plan. Results are immediates.

We use a standard edition without any customization

**Comment by molivasdat** [comment]
@ggdn Thanks for doing the additional investigation!

**Comment by callain** [comment]
We are experiencing the same issue for months in our project.

You have tagged this issue as `hacktoberfest`, how could we help ?

I have attached an explain plan of the query mentioned by @ggdn.

[explain.html.zip](https://github.com/liquibase/liquibase/files/5341792/explain.html.zip)

**Comment by molivasdat** [comment]
Hi @callain We tagged this with hacktoberfest so that anyone with a fix could find and create a PR to fix it. 

@nvoxland Do you have any suggestions on this one for @ggdn or @callain ?

**Comment by callain** [comment]
After doing some investigations, it seems that everything starts with this part of code.

StandardChangeLogHistoryService.java
```java
        Table changeLogTable = null;
        try {
            changeLogTable = SnapshotGeneratorFactory.getInstance().getDatabaseChangeLogTable(new SnapshotControl
                (database, false, Table.class, Column.class), database);
        } catch (LiquibaseException e) {
            throw new UnexpectedLiquibaseException(e);
        }
```

It get to this part in DatabaseSnapshot
```java
T object = chain.snapshot(example, this);
```

This is where it executes this SQL query as much as there are columns in the DATABASECHANGELOG table.

```sql
SELECT
    NULL            AS table_cat,
    atc.owner       AS table_schema,
    atc.owner,
    atc.table_name,
    atc.column_name,
    nullable,
    ac.validated    AS validated,
    ac.search_condition,
    ac.constraint_name
FROM
         all_tab_cols atc
    JOIN all_cons_columns  acc ON atc.owner = acc.owner
                                 AND atc.table_name = acc.table_name
                                 AND atc.column_name = acc.column_name
    JOIN all_constraints   ac ON atc.owner = ac.owner
                               AND atc.table_name = ac.table_name
                               AND acc.constraint_name = ac.constraint_name
WHERE
        atc.owner = 'MY_SCHEMA'
    AND atc.hidden_column = 'NO'
    AND ac.constraint_type = 'C'
    AND ac.search_condition IS NOT NULL
    AND atc.table_name = 'DATABASECHANGELOG'
```

I can reproduces it every time if I clean the sharedPool and the bufferCache before starting the app
```sql
ALTER SYSTEM FLUSH SHARED_POOL;
ALTER SYSTEM FLUSH BUFFER_CACHE;
```

What I understand is liquibase need to check the confirmity of the DATABASECHANGELOG table in a generic way.
You are doing this using a `Snapshot` concept I don't understand yet.
Looping on all DATABASECHANGELOG columns somewhere, but why ?

**Comment by molivasdat** [comment]
Hi @callain .  Thanks for the excellent review.  I'll get with the team and follow up with more info.

**Comment by nazarovkv** [comment]
Hi all, 

For Oracle suggest using prepared statement here: https://github.com/liquibase/liquibase/blob/5587a49248fbbf5cc12964b1a2fe098635f9c5ae/liquibase-core/src/main/java/liquibase/snapshot/JdbcDatabaseSnapshot.java#L892

**Comment by nazarovkv** [comment]
Another question: why for oracle global dictionary used? Why not use `USER_` tables?

**Comment by PascalSchumacher** [comment]
>Another question: why for oracle global dictionary used? Why not use USER_ tables?

Yes this change (if possible) would make it a lot faster if you have a lot of schemas/users sharing the database.

**Comment by syjer** [comment]
Hi, as we were suffering this issue, I did a small workaround project: https://github.com/syjer/liquibaseorausertable .

It basically implement a custom LockService/ChangeLogHistoryService with some of the problematic queries swapped out (that will go to the USER_ table). To be noted, this fix our use case, as we only use a small subset of liquibase capabilities.

**Comment by nazarovkv** [comment]
@syjer can you please share pom.xml example configuration?

**Comment by syjer** [comment]
@nazarovkv , basically, you include the dependency in your project (https://github.com/syjer/liquibaseorausertable#maven-coordinates) and it should work without any intervention (as the custom classes are automatically picked up by liquibase (https://github.com/syjer/liquibaseorausertable/tree/main/src/main/resources/META-INF/services)) . Currently we are using it in a few spring-boot projects where the migration is done by the application at boot time, so other use case are not really covered/tested.

**Comment by nazarovkv** [comment]
@syjer thanks for respond!
We are using liquibase both in spring boot application and on CI.
On CI liquibase invoked via liquibase-maven-plugin. 
I've tried to configure your lib as  dependency for liquibase-maven-plugin, but see no benefit for now. I guess I'm doing smth wrong.

**Comment by syjer** [comment]
@nazarovkv 

>liquibase-maven-plugin

When I got some time, I'll have a look what we need to do to make it work with the maven plugin. But I fear that if you  trigger any code path that do some snapshotting, it will be slow :(.

**Comment by nazarovkv** [comment]
@syjer well, I guess `dropAll` liquibase command requires snapshotting...
Our scenario on CI: prior to run unit tests, schema should be prepared. Hence first we do `dropAll` - in order to cleanup schema from previous builds, then do `updateTestingRollback` - in order to prepare schema for unit tests and test rollback scenarios.

**Comment by syjer** [comment]
@nazarovkv , yes you are right. When I got some additional time, I'll have a look at what it would mean to switch all the snapshotting related queries.

**Comment by nazarovkv** [comment]
@syjer that would be awesome! Looking forward to implementation! Thanks in advance!

**Comment by rodislav** [comment]
Hello, are there any news on this topic ?

**Comment by RichardBradley** [comment]
#2190 won't help with this, I think.

The problem you seem to have here is that the way that Liquibase is checking that the DATABASECHANGELOG table exists and has the correct columns is very slow on your Oracle (the "createSnapshot" code).

#2190 will still do most of this work, as `checkLiquibaseTables` is on by default. That PR will stop Liquibase from taking a "lock" in some cases, which is independent of the problem here.

I think the correct fix here is:

a) Determine if there is a faster way of inspecting the schema of a table in Oracle than the current approach and use it if so. Your idea to use `USER_ *` rather than `all_tab_cols` sounds promising. You will need to take care that any changes don't break other users of Oracle who may have different schema setups.

b) Double check whether Liquibase really needs to inspect the schema in this way, and consider simplifying the startup process to skip that, if it does not. This will require a bit more thought and investigation than (a), but may be a better end result. It seems to me that, if the goal is to handle schema mismatch in the Liquibase tables (DATABASECHANGELOG and others), perhaps we should just do a "SELECT" and handle errors + check the shape of the results, rather than inspecting the meta tables. If the Liquibase tables are not in the shape we expect, we will get a "no such column" error from the db, and can handle that.

(It is a bit surprising to me that Oracle will take so long just to tell you what columns are in its tables.)

GL

**Comment by kataggart** [comment]
@rs-dpatil what version of Liquibase are you using? We have done a few performance improvement fixes definitely since this was first reported, so want to first confirm that you have tried with a recent version? Thanks!

**Comment by rs-dpatil** [comment]
@kataggart I am using <liquibase.version>4.4.3</liquibase.version> with SpringBoot

**Comment by kataggart** [comment]
@rs-dpatil does your application allow you to test with a more current version?

**Comment by rs-dpatil** [comment]
I tried 4.8 version and it improved the performance a lot..!  ~ < 5 sec
